File: machine/cpp/rompack/format.cpp
/*
 * format.cpp - ROM pack utilities
 */

#include "format.h"
#include "common/endian.h"
#include "spec/bmsx/memory_map.h"
#include "spec/bmsx/rom_header.h"

namespace bmsx {
namespace {

void assertSectionRange(size_t offset, size_t length, size_t total, const char* label) {
	if (offset > total || length > total - offset) {
		throw BMSX_RUNTIME_ERROR(std::string("Invalid ROM ") + label + " range.");
	}
}

} // namespace

void writeCartRomHeader(u8* data, const CartRomHeader& header) {
	writeLE32(data + CART_ROM_HEADER_MAGIC_OFFSET, CART_ROM_MAGIC);
	writeLE32(data + CART_ROM_HEADER_SIZE_OFFSET, header.headerSize);
	writeLE32(data + CART_ROM_HEADER_MANIFEST_OFFSET, header.manifestOffset);
	writeLE32(data + CART_ROM_HEADER_MANIFEST_LENGTH_OFFSET, header.manifestLength);
	writeLE32(data + CART_ROM_HEADER_TOC_OFFSET, header.tocOffset);
	writeLE32(data + CART_ROM_HEADER_TOC_LENGTH_OFFSET, header.tocLength);
	writeLE32(data + CART_ROM_HEADER_DATA_OFFSET, header.dataOffset);
	writeLE32(data + CART_ROM_HEADER_DATA_LENGTH_OFFSET, header.dataLength);
	writeLE32(data + BMSX_ROM_HEADER_BLUA32_IMAGE_OFFSET, header.blua32ImageOffset);
	writeLE32(
		data + BMSX_ROM_HEADER_BLUA32_IMAGE_BYTE_COUNT_OFFSET,
		header.blua32ImageByteCount
	);
	writeLE32(
		data + BMSX_ROM_HEADER_BLUA32_STARTUP_FUNCTION_ADDRESS_OFFSET,
		header.blua32StartupFunctionAddress
	);
	writeLE32(
		data + BMSX_ROM_HEADER_BLUA32_IRQ_FUNCTION_ADDRESS_OFFSET,
		header.blua32IrqFunctionAddress
	);
	writeLE32(
		data + BMSX_ROM_HEADER_BLUA32_EXCEPTION_FUNCTION_ADDRESS_OFFSET,
		header.blua32ExceptionFunctionAddress
	);
	writeLE32(
		data + BMSX_ROM_HEADER_BLUA32_STATIC_LAYOUT_TOKEN_LO_OFFSET,
		header.blua32StaticLayoutTokenLo
	);
	writeLE32(
		data + BMSX_ROM_HEADER_BLUA32_STATIC_LAYOUT_TOKEN_HI_OFFSET,
		header.blua32StaticLayoutTokenHi
	);
	writeLE32(
		data + CART_ROM_HEADER_BLUA32_DIAGNOSTIC_DIRECTORY_OFFSET,
		header.blua32DiagnosticDirectoryOffset
	);
	writeLE32(data + CART_ROM_HEADER_METADATA_OFFSET, header.metadataOffset);
	writeLE32(data + CART_ROM_HEADER_METADATA_LENGTH_OFFSET, header.metadataLength);
	writeLE32(data + CART_ROM_HEADER_RESERVED_1_OFFSET, 0u);
	writeLE32(data + CART_ROM_HEADER_CARTRIDGE_BOARD_OFFSET, header.cartridgeBoardWord);
	writeLE32(data + CART_ROM_HEADER_CARTRIDGE_RAM_BYTES_OFFSET, header.cartridgeRamByteCount);
}

CartRomHeader parseCartHeader(const u8* data, size_t size) {
	if (size < CART_ROM_HEADER_SIZE) {
		throw BMSX_RUNTIME_ERROR("ROM payload is too small for cart header.");
	}
	if (readLE32(data + CART_ROM_HEADER_MAGIC_OFFSET) != CART_ROM_MAGIC) {
		throw BMSX_RUNTIME_ERROR("Invalid ROM cart header.");
	}
	CartRomHeader header{};
	header.headerSize = readLE32(data + CART_ROM_HEADER_SIZE_OFFSET);
	if (header.headerSize < CART_ROM_HEADER_SIZE) {
		throw BMSX_RUNTIME_ERROR("ROM header size is too small.");
	}
	if (header.headerSize > size) {
		throw BMSX_RUNTIME_ERROR("ROM header size exceeds payload length.");
	}
	header.manifestOffset = readLE32(data + CART_ROM_HEADER_MANIFEST_OFFSET);
	header.manifestLength = readLE32(data + CART_ROM_HEADER_MANIFEST_LENGTH_OFFSET);
	header.tocOffset = readLE32(data + CART_ROM_HEADER_TOC_OFFSET);
	header.tocLength = readLE32(data + CART_ROM_HEADER_TOC_LENGTH_OFFSET);
	header.dataOffset = readLE32(data + CART_ROM_HEADER_DATA_OFFSET);
	header.dataLength = readLE32(data + CART_ROM_HEADER_DATA_LENGTH_OFFSET);
	header.blua32ImageOffset = readLE32(data + BMSX_ROM_HEADER_BLUA32_IMAGE_OFFSET);
	header.blua32ImageByteCount = readLE32(
		data + BMSX_ROM_HEADER_BLUA32_IMAGE_BYTE_COUNT_OFFSET
	);
	header.blua32StartupFunctionAddress = readLE32(
		data + BMSX_ROM_HEADER_BLUA32_STARTUP_FUNCTION_ADDRESS_OFFSET
	);
	header.blua32IrqFunctionAddress = readLE32(
		data + BMSX_ROM_HEADER_BLUA32_IRQ_FUNCTION_ADDRESS_OFFSET
	);
	header.blua32ExceptionFunctionAddress = readLE32(
		data + BMSX_ROM_HEADER_BLUA32_EXCEPTION_FUNCTION_ADDRESS_OFFSET
	);
	header.blua32StaticLayoutTokenLo = readLE32(
		data + BMSX_ROM_HEADER_BLUA32_STATIC_LAYOUT_TOKEN_LO_OFFSET
	);
	header.blua32StaticLayoutTokenHi = readLE32(
		data + BMSX_ROM_HEADER_BLUA32_STATIC_LAYOUT_TOKEN_HI_OFFSET
	);
	header.blua32DiagnosticDirectoryOffset = readLE32(
		data + CART_ROM_HEADER_BLUA32_DIAGNOSTIC_DIRECTORY_OFFSET
	);
	header.metadataOffset = readLE32(data + CART_ROM_HEADER_METADATA_OFFSET);
	header.metadataLength = readLE32(data + CART_ROM_HEADER_METADATA_LENGTH_OFFSET);
	header.cartridgeBoardWord = readLE32(data + CART_ROM_HEADER_CARTRIDGE_BOARD_OFFSET);
	header.cartridgeRamByteCount = readLE32(data + CART_ROM_HEADER_CARTRIDGE_RAM_BYTES_OFFSET);
	if (header.cartridgeRamByteCount > CART_RAM_SIZE) {
		throw BMSX_RUNTIME_ERROR("Cartridge RAM byte count exceeds the socket aperture.");
	}

	assertSectionRange(static_cast<size_t>(header.manifestOffset), static_cast<size_t>(header.manifestLength), size, "manifest");
	assertSectionRange(static_cast<size_t>(header.tocOffset), static_cast<size_t>(header.tocLength), size, "toc");
	assertSectionRange(static_cast<size_t>(header.dataOffset), static_cast<size_t>(header.dataLength), size, "data");
	if (header.metadataLength > 0) {
		assertSectionRange(static_cast<size_t>(header.metadataOffset), static_cast<size_t>(header.metadataLength), size, "metadata");
	}
	return header;
}

} // namespace bmsx

File: machine/ts/rompack/format.ts
import {
	BMSX_ROM_HEADER_BLUA32_EXCEPTION_FUNCTION_ADDRESS_OFFSET,
	BMSX_ROM_HEADER_BLUA32_IMAGE_BYTE_COUNT_OFFSET,
	BMSX_ROM_HEADER_BLUA32_IMAGE_OFFSET,
	BMSX_ROM_HEADER_BLUA32_IRQ_FUNCTION_ADDRESS_OFFSET,
	BMSX_ROM_HEADER_BLUA32_STARTUP_FUNCTION_ADDRESS_OFFSET,
	BMSX_ROM_HEADER_BLUA32_STATIC_LAYOUT_TOKEN_HI_OFFSET,
	BMSX_ROM_HEADER_BLUA32_STATIC_LAYOUT_TOKEN_LO_OFFSET,
} from '../spec/bmsx/rom_header';
import {
	CART_ROM_HEADER_BLUA32_DIAGNOSTIC_DIRECTORY_OFFSET,
	CART_ROM_HEADER_DATA_LENGTH_OFFSET,
	CART_ROM_HEADER_DATA_OFFSET,
	CART_ROM_HEADER_MANIFEST_LENGTH_OFFSET,
	CART_ROM_HEADER_MANIFEST_OFFSET,
	CART_ROM_HEADER_MAGIC_OFFSET,
	CART_ROM_HEADER_METADATA_LENGTH_OFFSET,
	CART_ROM_HEADER_METADATA_OFFSET,
	CART_ROM_HEADER_SIZE,
	CART_ROM_HEADER_SIZE_OFFSET,
	CART_ROM_HEADER_TOC_LENGTH_OFFSET,
	CART_ROM_HEADER_TOC_OFFSET,
	CART_ROM_MAGIC,
} from '../spec/bmsx/rom_package';
import { formatNumberAsHex } from '../common/byte_hex_string';

export type CartRomHeader = {
	headerSize: number;
	manifestOffset: number;
	manifestLength: number;
	tocOffset: number;
	tocLength: number;
	dataOffset: number;
	dataLength: number;
	blua32ImageOffset: number;
	blua32ImageByteCount: number;
	blua32StartupFunctionAddress: number;
	blua32IrqFunctionAddress: number;
	blua32ExceptionFunctionAddress: number;
	blua32StaticLayoutTokenLo: number;
	blua32StaticLayoutTokenHi: number;
	blua32DiagnosticDirectoryOffset: number;
	metadataOffset: number;
	metadataLength: number;
};

function assertRomSectionRange(offset: number, length: number, total: number, label: string): void {
	if (offset + length > total) {
		throw new Error(`Invalid ROM ${label} range: offset=${formatNumberAsHex(offset)} len=${formatNumberAsHex(length)} total=${formatNumberAsHex(total)}.`);
	}
}

export function parseCartHeader(payload: Uint8Array): CartRomHeader {
	if (payload.byteLength < CART_ROM_HEADER_SIZE) {
		throw new Error('ROM payload is too small for cart header.');
	}
	const view = new DataView(payload.buffer, payload.byteOffset, CART_ROM_HEADER_SIZE);
	if (view.getUint32(CART_ROM_HEADER_MAGIC_OFFSET, true) !== CART_ROM_MAGIC) {
		throw new Error('Invalid ROM cart header.');
	}
	const headerSize = view.getUint32(CART_ROM_HEADER_SIZE_OFFSET, true);
	if (headerSize !== CART_ROM_HEADER_SIZE) {
		throw new Error(`ROM header size ${headerSize} does not match the current ${CART_ROM_HEADER_SIZE}-byte format.`);
	}
	const manifestOffset = view.getUint32(CART_ROM_HEADER_MANIFEST_OFFSET, true);
	const manifestLength = view.getUint32(CART_ROM_HEADER_MANIFEST_LENGTH_OFFSET, true);
	const tocOffset = view.getUint32(CART_ROM_HEADER_TOC_OFFSET, true);
	const tocLength = view.getUint32(CART_ROM_HEADER_TOC_LENGTH_OFFSET, true);
	const dataOffset = view.getUint32(CART_ROM_HEADER_DATA_OFFSET, true);
	const dataLength = view.getUint32(CART_ROM_HEADER_DATA_LENGTH_OFFSET, true);
	const metadataOffset = view.getUint32(CART_ROM_HEADER_METADATA_OFFSET, true);
	const metadataLength = view.getUint32(CART_ROM_HEADER_METADATA_LENGTH_OFFSET, true);

	assertRomSectionRange(manifestOffset, manifestLength, payload.byteLength, 'manifest');
	assertRomSectionRange(tocOffset, tocLength, payload.byteLength, 'toc');
	assertRomSectionRange(dataOffset, dataLength, payload.byteLength, 'data');
	if (metadataLength > 0) {
		assertRomSectionRange(metadataOffset, metadataLength, payload.byteLength, 'metadata');
	}

	return {
		headerSize,
		manifestOffset,
		manifestLength,
		tocOffset,
		tocLength,
		dataOffset,
		dataLength,
		blua32ImageOffset: view.getUint32(BMSX_ROM_HEADER_BLUA32_IMAGE_OFFSET, true),
		blua32ImageByteCount: view.getUint32(BMSX_ROM_HEADER_BLUA32_IMAGE_BYTE_COUNT_OFFSET, true),
		blua32StartupFunctionAddress: view.getUint32(BMSX_ROM_HEADER_BLUA32_STARTUP_FUNCTION_ADDRESS_OFFSET, true),
		blua32IrqFunctionAddress: view.getUint32(BMSX_ROM_HEADER_BLUA32_IRQ_FUNCTION_ADDRESS_OFFSET, true),
		blua32ExceptionFunctionAddress: view.getUint32(BMSX_ROM_HEADER_BLUA32_EXCEPTION_FUNCTION_ADDRESS_OFFSET, true),
		blua32StaticLayoutTokenLo: view.getUint32(BMSX_ROM_HEADER_BLUA32_STATIC_LAYOUT_TOKEN_LO_OFFSET, true),
		blua32StaticLayoutTokenHi: view.getUint32(BMSX_ROM_HEADER_BLUA32_STATIC_LAYOUT_TOKEN_HI_OFFSET, true),
		blua32DiagnosticDirectoryOffset: view.getUint32(CART_ROM_HEADER_BLUA32_DIAGNOSTIC_DIRECTORY_OFFSET, true),
		metadataOffset,
		metadataLength,
	};
}
